src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,296 @@
+                    trade, order, instance2, order_type[-1]
+                )
+                diff_surplus = winning_surplus - qmdo_surplus
+                (percent_deviation, diff_in_eth) = self.get_conversions(
+                    diff_surplus,
+                    trade,
+                    order_type[-1],
+                    bucket_response["tokens"],
+                    order,
+                )
+                if percent_deviation < 0.1 and diff_in_eth < 0.002:
+                    self.print_logs(
+                        settlement_hash, order_id, diff_in_eth, percent_deviation
+                    )
+        return bucket_response
+
+    def print_logs(self, settlement_hash, order_id, diff_in_eth, percent_deviation):
+        self.logger.info(
+            "Settlement Hash: %s\nFor order: %s\nAbsolute ETH Difference: %s\nRelative Percent Difference: %s\n",
+            settlement_hash,
+            order_id,
+            str(format(diff_in_eth, ".5f")),
+            str(format(percent_deviation, ".4f")),
+        )
+
+    def get_conversions(self, diff_surplus, trade, order_type, tokens, order):
+        """
+        calcuate order flag condition values,
+        (relative % deviation, absolute ETH difference) based on order type
+        """
+        # implies a sell order
+        if order_type == "0":
+            percent_deviation = (diff_surplus * 100) / int(trade["buyAmount"])
+            buy_token = order["buy_token"]
+            diff_in_eth = (
+                tokens[buy_token]["external_price"] / (pow(10, 18)) * (diff_surplus)
+            )
+        # implies a buy order
+        elif order_type == "1":
+            percent_deviation = (diff_surplus * 100) / int(trade["sellAmount"])
+            sell_token = order["buy_token"]
+            diff_in_eth = (
+                tokens[sell_token]["external_price"] / (pow(10, 18)) * (diff_surplus)
+            )
+
+        return percent_deviation, diff_in_eth
+
+    def get_surplus(
+        self, trade, sell_token_clearing_price, buy_token_clearing_price, order_type
+    ):
+        """
+        Function calculates surplus using clearing prices,
+        for winning order and from quasimodo
+        """
+        # implies sell order
+        if order_type == "0":
+            executed_volume = int(
+                Fraction(trade["executedAmount"])
+                * Fraction(sell_token_clearing_price)
+                // Fraction(buy_token_clearing_price)
+            )
+            Surplus = executed_volume - int(trade["buyAmount"])
+        # implies buy order
+        elif order_type == "1":
+            executed_volume = int(
+                Fraction(trade["executedAmount"])
+                * Fraction(buy_token_clearing_price)
+                // Fraction(sell_token_clearing_price)
+            )
+            Surplus = int(trade["sellAmount"]) - executed_volume
+        return Surplus
+
+
+def get_logger() -> logging.Logger:
```

Comment:
I am tempted to say that this code appears as is in the offchain part. If that is the case, then we should avoid repetition.

Reply:
are you referencing the get_logger function?